File: src/carbonfactor_parser/persistence/parsed_factor_persistence_writer.py
"""Parsed emission factor persistence writer boundary."""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
from enum import Enum
import hashlib
import json
from typing import TYPE_CHECKING, Mapping, Sequence

from carbonfactor_parser.parsers.raw_record import (
    ParsedRawRecord,
    ParsedRawRecordPayload,
    validate_parsed_raw_record_payload,
)
from carbonfactor_parser.persistence.postgresql_schema_catalog import SourceFamily
from carbonfactor_parser.persistence.source_document_mapping import (
    DRY_RUN_TIMESTAMP_LABEL,
)
from carbonfactor_parser.persistence.source_family_repository import (
    SourceFamilyDetailRecord,
    SourceFamilyMasterRecord,
    SourceFamilyRepository,
    SourceFamilyRepositoryIssue,
    SourceFamilyRepositoryPersistStatus,
    validate_source_family_repository_inputs,
)

if TYPE_CHECKING:
    from carbonfactor_parser.parsers.normalized_output_row_contract import (
        ParserNormalizedOutputBatch,
        ParserNormalizedOutputRow,
    )
    ParsedFactorOutput = ParsedRawRecordPayload | ParserNormalizedOutputBatch
else:
    ParsedFactorOutput = object


class ParsedFactorPersistenceStatus(str, Enum):
    """Status for parsed factor persistence writer outcomes."""

    DECLARED = "declared"
    FAILED_VALIDATION = "failed_validation"
    NO_RECORDS = "no_records"


@dataclass(frozen=True)
class ParsedFactorPersistenceIssue:
    """Validation or repository issue from parsed factor persistence mapping."""

    code: str
    message: str
    field_name: str
    severity: str = "error"


@dataclass(frozen=True)
class ParsedFactorPersistenceCommand:
    """Source-family master/detail records ready for repository persistence."""

    master_records: tuple[SourceFamilyMasterRecord, ...]
    detail_records: tuple[SourceFamilyDetailRecord, ...]
    skipped_duplicate_count: int = 0
    issues: tuple[ParsedFactorPersistenceIssue, ...] = ()


@dataclass(frozen=True)
class ParsedFactorPersistenceWriterResult:
    """Result of building and submitting parsed factor persistence commands."""

    provider_name: str
    status: ParsedFactorPersistenceStatus
    attempted_master_count: int
    attempted_detail_count: int
    persisted_master_count: int
    persisted_detail_count: int
    skipped_duplicate_count: int = 0
    issues: tuple[ParsedFactorPersistenceIssue, ...] = ()
    command: ParsedFactorPersistenceCommand | None = None


_SOURCE_FAMILY_ALIASES: Mapping[str, SourceFamily] = {
    "ghg": SourceFamily.GHG,
    "ghg_protocol": SourceFamily.GHG,
    "defra": SourceFamily.DEFRA,
    "defra_desnz": SourceFamily.DEFRA,
    "desnz": SourceFamily.DEFRA,
    "ipcc": SourceFamily.IPCC,
    "ipcc_efdb": SourceFamily.IPCC,
}


def build_parsed_factor_persistence_command(
    parsed_output: ParsedFactorOutput,
    *,
    source_document_id: str | None = None,
    lifecycle_status: str = "active",
    timestamp_label: str = DRY_RUN_TIMESTAMP_LABEL,
) -> ParsedFactorPersistenceCommand:
    """Map parsed factor output into source-family master/detail records."""

    rows, shape_issues = _rows_from_output(parsed_output)
    if not rows and not shape_issues:
        return ParsedFactorPersistenceCommand(
            master_records=(),
            detail_records=(),
            issues=(
                ParsedFactorPersistenceIssue(
                    code="PARSED_FACTOR_PERSISTENCE_NO_RECORDS",
                    message="parsed output must include records before persistence.",
                    field_name="records",
                    severity="warning",
                ),
            ),
        )

    issues: list[ParsedFactorPersistenceIssue] = list(shape_issues)
    masters: dict[tuple[SourceFamily, str], SourceFamilyMasterRecord] = {}
    details: dict[tuple[SourceFamily, str, str], SourceFamilyDetailRecord] = {}
    skipped_duplicate_count = 0

    for position, row in enumerate(rows, start=1):
        mapped = _map_row(
            row,
            position=position,
            source_document_id=source_document_id,
            lifecycle_status=lifecycle_status,
            timestamp_label=timestamp_label,
        )
        issues.extend(mapped.issues)
        if mapped.master_record is None or mapped.detail_record is None:
            continue

        master_key = (
            mapped.master_record.source_family,
            mapped.master_record.source_family_master_id,
        )
        existing_master = masters.get(master_key)
        if existing_master is None:
            masters[master_key] = mapped.master_record
        elif existing_master == mapped.master_record:
            skipped_duplicate_count += 1
        else:
            issues.append(
                ParsedFactorPersistenceIssue(
                    code="PARSED_FACTOR_PERSISTENCE_DUPLICATE_MASTER_CONFLICT",
                    message=(
                        "duplicate source-family master identity maps to "
                        "different record content."
                    ),
                    field_name=f"records[{position}].source_family_master_id",
                ),
            )

        detail_key = (
            mapped.detail_record.source_family,
            mapped.detail_record.source_family_master_id,
            mapped.detail_record.detail_external_key,
        )
        existing_detail = details.get(detail_key)
        if existing_detail is None:
            details[detail_key] = mapped.detail_record
        elif existing_detail == mapped.detail_record:
            skipped_duplicate_count += 1
        else:
            issues.append(
                ParsedFactorPersistenceIssue(
                    code="PARSED_FACTOR_PERSISTENCE_DUPLICATE_DETAIL_CONFLICT",
                    message=(
                        "duplicate factor identity maps to different detail "
                        "record content."
                    ),
                    field_name=f"records[{position}].detail_external_key",
                ),
            )

    command = ParsedFactorPersistenceCommand(
        master_records=tuple(masters.values()),
        detail_records=tuple(details.values()),
        skipped_duplicate_count=skipped_duplicate_count,
        issues=tuple(issues),
    )
    repository_validation = validate_source_family_repository_inputs(
        provider_name="parsed_factor_persistence_command",
        master_records=command.master_records,
        detail_records=command.detail_records,
    )
    if repository_validation.issues:
        return ParsedFactorPersistenceCommand(
            master_records=command.master_records,
            detail_records=command.detail_records,
            skipped_duplicate_count=command.skipped_duplicate_count,
            issues=(
                *command.issues,
                *tuple(
                    _from_repository_issue(issue)
                    for issue in repository_validation.issues
                ),
            ),
        )

    return command


def persist_parsed_factor_records(
    parsed_output: ParsedFactorOutput,
    repository: SourceFamilyRepository,
    *,
    source_document_id: str | None = None,
    lifecycle_status: str = "active",
    timestamp_label: str = DRY_RUN_TIMESTAMP_LABEL,
) -> ParsedFactorPersistenceWriterResult:
    """Build parsed factor records and submit them to a repository protocol."""

    command = build_parsed_factor_persistence_command(
        parsed_output,
        source_document_id=source_document_id,
        lifecycle_status=lifecycle_status,
        timestamp_label=timestamp_label,
    )
    if command.issues:
        status = (
            ParsedFactorPersistenceStatus.NO_RECORDS
            if _only_no_records(command.issues)
            else ParsedFactorPersistenceStatus.FAILED_VALIDATION
        )
        return ParsedFactorPersistenceWriterResult(
            provider_name=repository.provider_name,
            status=status,
            attempted_master_count=len(command.master_records),
            attempted_detail_count=len(command.detail_records),
            persisted_master_count=0,
            persisted_detail_count=0,
            skipped_duplicate_count=command.skipped_duplicate_count,
            issues=command.issues,
            command=command,
        )

    repository_result = repository.persist_source_family_records(
        command.master_records,
        command.detail_records,
    )
    repository_issues = tuple(
        _from_repository_issue(issue) for issue in repository_result.issues
    )
    status = (
        ParsedFactorPersistenceStatus.DECLARED
        if repository_result.status is SourceFamilyRepositoryPersistStatus.DECLARED
        else ParsedFactorPersistenceStatus.FAILED_VALIDATION
    )
    return ParsedFactorPersistenceWriterResult(
        provider_name=repository_result.provider_name,
        status=status,
        attempted_master_count=len(command.master_records),
        attempted_detail_count=len(command.detail_records),
        persisted_master_count=repository_result.persisted_master_count,
        persisted_detail_count=repository_result.persisted_detail_count,
        skipped_duplicate_count=command.skipped_duplicate_count,
        issues=repository_issues,
        command=command,
    )


@dataclass(frozen=True)
class _PersistenceRow:
    source_family: str
    source_id: str
    row_id: str
    fields: Mapping[str, object]
    artifact_reference: str | None
    artifact_checksum_sha256: str | None
    source_row_number: int | None


@dataclass(frozen=True)
class _MappedRow:
    master_record: SourceFamilyMasterRecord | None
    detail_record: SourceFamilyDetailRecord | None
    issues: tuple[ParsedFactorPersistenceIssue, ...] = ()


def _rows_from_output(
    parsed_output: ParsedFactorOutput,
) -> tuple[tuple[_PersistenceRow, ...], tuple[ParsedFactorPersistenceIssue, ...]]:
    from carbonfactor_parser.parsers.normalized_output_row_contract import (
        validate_parser_normalized_output_batch,
    )

    if _looks_like_raw_record_payload(parsed_output):
        validation = validate_parsed_raw_record_payload(parsed_output)
        issues = tuple(
            ParsedFactorPersistenceIssue(
                code=issue.code,
                message=issue.message,
                field_name=issue.field_name,
                severity=issue.severity,
            )
            for issue in validation.issues
        )
        return (
            tuple(_row_from_raw_record(record) for record in parsed_output.records),
            issues,
        )

    if _looks_like_normalized_output_batch(parsed_output):
        validation = validate_parser_normalized_output_batch(parsed_output)
        issues = tuple(
            ParsedFactorPersistenceIssue(
                code=issue.code,
                message=issue.message,
                field_name=issue.field_name,
                severity=issue.severity,
            )
            for issue in validation.issues
        )
        return (
            tuple(_row_from_normalized_row(row) for row in parsed_output.rows),
            issues,
        )

    return (
        (),
        (
            ParsedFactorPersistenceIssue(
                code="PARSED_FACTOR_PERSISTENCE_INVALID_OUTPUT",
                message=(
                    "parsed_output must be ParsedRawRecordPayload or "
                    "ParserNormalizedOutputBatch."
                ),
                field_name="parsed_output",
            ),
        ),
    )


def _looks_like_raw_record_payload(value: object) -> bool:
    return hasattr(value, "records") and all(
        hasattr(record, "raw_fields") for record in getattr(value, "records", ())
    )


def _looks_like_normalized_output_batch(value: object) -> bool:
    return hasattr(value, "rows") and all(
        hasattr(row, "normalized_fields") for row in getattr(value, "rows", ())
    )


def _row_from_raw_record(record: ParsedRawRecord) -> _PersistenceRow:
    return _PersistenceRow(
        source_family=record.source_family,
        source_id=record.source_id,
        row_id=f"record-{record.record_index}",
        fields=dict(record.raw_fields),
        artifact_reference=_text_or_none(
            (record.source_context or {}).get("artifact_reference")
        ),
        artifact_checksum_sha256=_text_or_none(
            (record.source_context or {}).get("checksum_sha256")
        ),
        source_row_number=record.row_number,
    )


def _row_from_normalized_row(row: "ParserNormalizedOutputRow") -> _PersistenceRow:
    return _PersistenceRow(
        source_family=row.source_family,
        source_id=row.source_key,
        row_id=row.row_id,
        fields=dict(row.normalized_fields),
        artifact_reference=row.artifact_reference,
        artifact_checksum_sha256=_text_or_none(
            _field(row.normalized_fields, "source_checksum_sha256", "checksum_sha256")
        ),
        source_row_number=row.source_row_number,
    )


def _map_row(
    row: _PersistenceRow,
    *,
    position: int,
    source_document_id: str | None,
    lifecycle_status: str,
    timestamp_label: str,
) -> _MappedRow:
    issues: list[ParsedFactorPersistenceIssue] = []
    source_family = _source_family(row.source_family)
    if source_family is None:
        issues.append(
            ParsedFactorPersistenceIssue(
                code="PARSED_FACTOR_PERSISTENCE_UNSUPPORTED_SOURCE_FAMILY",
                message="source_family must map to GHG, DEFRA/DESNZ, or IPCC.",
                field_name=f"records[{position}].source_family",
            ),
        )
        return _MappedRow(None, None, tuple(issues))

    resolved_source_document_id = _source_document_id(row, source_document_id)
    required_values = {
        "source_document_id": resolved_source_document_id,
        "factor_value": _field(row.fields, "factor_value", "value"),
        "factor_unit": _field(row.fields, "factor_unit", "unit"),
    }
    for field_name, value in required_values.items():
        if _text_or_none(value) is None:
            issues.append(
                ParsedFactorPersistenceIssue(
                    code="PARSED_FACTOR_PERSISTENCE_MISSING_REQUIRED_FIELD",
                    message="parsed factor persistence requires a non-empty value.",
                    field_name=f"records[{position}].{field_name}",
                ),
            )

    if issues:
        return _MappedRow(None, None, tuple(issues))

    master_external_key = _text_or_none(
        _field(row.fields, "master_external_key")
    ) or _default_master_external_key(row)
    detail_external_key = _text_or_none(
        _field(row.fields, "detail_external_key")
    ) or _default_detail_external_key(row)
    master_id = _text_or_none(
        _field(row.fields, "source_family_master_id")
    ) or (
        f"{source_family.value}_master_"
        f"{_stable_digest(source_family.value, master_external_key)[:16]}"
    )
    detail_id = _text_or_none(
        _field(row.fields, "source_family_detail_id")
    ) or (
        f"{source_family.value}_detail_"
        f"{_stable_digest(source_family.value, master_id, detail_external_key)[:16]}"
    )
    source_year = _int_or_none(_field(row.fields, "source_year", "reporting_year"))
    if source_year is None:
        issues.append(
            ParsedFactorPersistenceIssue(
                code="PARSED_FACTOR_PERSISTENCE_MISSING_REQUIRED_FIELD",
                message="parsed factor persistence requires a source_year value.",
                field_name=f"records[{position}].source_year",
            ),
        )
        return _MappedRow(None, None, tuple(issues))
    source_version = _text_or_none(_field(row.fields, "source_version")) or "unknown"
    artifact_checksum_sha256 = row.artifact_checksum_sha256 or _text_or_none(
        _field(row.fields, "provenance_checksum_value", "source_checksum_sha256")
    )

    master_record = SourceFamilyMasterRecord(
        source_family=source_family,
        source_family_master_id=master_id,
        source_year=source_year,
        source_version=source_version,
        source_release=_text_or_none(_field(row.fields, "source_release")),
        source_document_id=resolved_source_document_id or "",
        ingestion_run_id=_text_or_none(_field(row.fields, "ingestion_run_id")),
        run_id=_text_or_none(_field(row.fields, "run_id")),
        master_external_key=master_external_key,
        status=lifecycle_status,
        artifact_reference=row.artifact_reference,
        artifact_checksum_sha256=artifact_checksum_sha256,
        archive_reference=_text_or_none(_field(row.fields, "archive_reference")),
        archive_checksum_sha256=_text_or_none(
            _field(row.fields, "archive_checksum_sha256")
        ),
        effective_from=_text_or_none(_field(row.fields, "effective_from")),
        effective_to=_text_or_none(_field(row.fields, "effective_to")),
        record_checksum_sha256=_record_checksum(
            "master",
            source_family.value,
            resolved_source_document_id,
            master_external_key,
            lifecycle_status,
        ),
        metadata={},
        created_at=timestamp_label,
        updated_at=timestamp_label,
    )
    detail_record = SourceFamilyDetailRecord(
        source_family=source_family,
        source_family_detail_id=detail_id,
        source_family_master_id=master_id,
        detail_external_key=detail_external_key,
        source_row_number=row.source_row_number,
        factor_id=_text_or_none(_field(row.fields, "factor_id")),
        factor_name=_text_or_none(_field(row.fields, "factor_name")),
        factor_value=_text_or_none(required_values["factor_value"]) or "",
        factor_unit=_text_or_none(required_values["factor_unit"]) or "",
        status=lifecycle_status,
        record_checksum_sha256=_record_checksum(
            "detail",
            source_family.value,
            master_id,
            detail_external_key,
            required_values["factor_value"],
            required_values["factor_unit"],
        ),
        raw_fields=dict(row.fields),
        normalized_fields=dict(row.fields),
        created_at=timestamp_label,
        updated_at=timestamp_label,
    )
    return _MappedRow(master_record, detail_record)


def _source_family(value: str) -> SourceFamily | None:
    if not isinstance(value, str):
        return None
    return _SOURCE_FAMILY_ALIASES.get(value.strip().lower())


def _source_document_id(
    row: _PersistenceRow,
    explicit_source_document_id: str | None,
) -> str | None:
    explicit = _text_or_none(explicit_source_document_id)
    if explicit is not None:
        return explicit
    field_value = _text_or_none(_field(row.fields, "source_document_id"))
    if field_value is not None:
        return field_value
    artifact_reference = _text_or_none(
        _field(row.fields, "provenance_artifact_reference", "artifact_reference")
    ) or row.artifact_reference
    checksum = _text_or_none(
        _field(row.fields, "provenance_checksum_value", "source_checksum_sha256")
    )
    if artifact_reference is None and checksum is None:
        return None
    digest = _stable_digest(
        row.source_family,
        row.source_id,
        artifact_reference,
        checksum,
    )
    return f"source_document_{digest[:24]}"


def _default_master_external_key(row: _PersistenceRow) -> str:
    source_year = _text_or_none(_field(row.fields, "source_year")) or "unknown-year"
    source_version = (
        _text_or_none(_field(row.fields, "source_version")) or "unknown-version"
    )
    factor_id = _text_or_none(_field(row.fields, "factor_id")) or row.row_id
    return f"{source_year}:{source_version}:{factor_id}"


def _default_detail_external_key(row: _PersistenceRow) -> str:
    factor_id = _text_or_none(_field(row.fields, "factor_id")) or row.row_id
    factor_unit = (
        _text_or_none(_field(row.fields, "factor_unit", "unit")) or "unknown-unit"
    )
    gas = _text_or_none(_field(row.fields, "greenhouse_gas", "gas"))
    if gas is None:
        return f"{factor_id}:{factor_unit}"
    return f"{factor_id}:{factor_unit}:{gas}"


def _field(fields: Mapping[str, object], *names: str) -> object | None:
    for name in names:
        if name in fields:
            return fields[name]
    return None


def _text_or_none(value: object | None) -> str | None:
    if value is None:
        return None
    text = str(value).strip()
    return text or None


def _int_or_none(value: object | None) -> int | None:
    if value is None or isinstance(value, bool):
        return None
    if isinstance(value, int):
        return value
    text = _text_or_none(value)
    if text is None:
        return None
    try:
        return int(text)
    except ValueError:
        return None


def _record_checksum(*values: object) -> str:
    return _stable_digest(*values)


def _stable_digest(*values: object) -> str:
    payload = json.dumps(
        [_json_safe(value) for value in values],
        sort_keys=True,
        separators=(",", ":"),
    )
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def _json_safe(value: object) -> object:
    if isinstance(value, Decimal):
        return str(value)
    if isinstance(value, Mapping):
        return {
            str(key): _json_safe(item)
            for key, item in sorted(value.items(), key=lambda pair: str(pair[0]))
        }
    if isinstance(value, Sequence) and not isinstance(value, str):
        return [_json_safe(item) for item in value]
    return value


def _from_repository_issue(
    issue: SourceFamilyRepositoryIssue,
) -> ParsedFactorPersistenceIssue:
    return ParsedFactorPersistenceIssue(
        code=issue.code,
        message=issue.message,
        field_name=issue.field_name,
        severity=issue.severity,
    )


def _only_no_records(issues: tuple[ParsedFactorPersistenceIssue, ...]) -> bool:
    return tuple(issue.code for issue in issues) == (
        "PARSED_FACTOR_PERSISTENCE_NO_RECORDS",
    )

File: src/dotnet/CarbonOps.Parser.Contracts/ParsedFactorPersistenceWriter.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace CarbonOps.Parser.Contracts;

public static class ParsedFactorPersistenceWriter
{
    public const string DefaultTimestampLabel = "dry_run_timestamp_unavailable";

    public static ParsedFactorPersistenceCommand BuildCommand(
        ParserNormalizedOutputBatch parsedOutput,
        string? sourceDocumentId = null,
        string lifecycleStatus = "active",
        string timestampLabel = DefaultTimestampLabel)
    {
        if (parsedOutput is null)
        {
            return new ParsedFactorPersistenceCommand(
                [],
                [],
                issues:
                [
                    new ParsedFactorPersistenceIssue(
                        "PARSED_FACTOR_PERSISTENCE_INVALID_OUTPUT",
                        "parsed output must be a ParserNormalizedOutputBatch.",
                        "parsedOutput"),
                ]);
        }

        if (parsedOutput.RowCount == 0)
        {
            return new ParsedFactorPersistenceCommand(
                [],
                [],
                issues:
                [
                    new ParsedFactorPersistenceIssue(
                        "PARSED_FACTOR_PERSISTENCE_NO_RECORDS",
                        "parsed output must include records before persistence.",
                        "Rows",
                        "warning"),
                ]);
        }

        var issues = new List<ParsedFactorPersistenceIssue>();
        var masters = new Dictionary<(SourceFamily SourceFamily, string MasterId), SourceFamilyMasterRecord>();
        var details = new Dictionary<(SourceFamily SourceFamily, string MasterId, string DetailExternalKey), SourceFamilyDetailRecord>();
        var skippedDuplicateCount = 0;

        for (var index = 0; index < parsedOutput.Rows.Count; index++)
        {
            var row = parsedOutput.Rows[index];
            if (row is null)
            {
                issues.Add(new ParsedFactorPersistenceIssue(
                    "PARSED_FACTOR_PERSISTENCE_INVALID_NORMALIZED_ROW",
                    "ParserNormalizedOutputRow is required.",
                    $"Rows[{index}]"));
                continue;
            }

            AppendRowValidationIssues(row, index, issues);

            var mapped = MapRow(
                row,
                index,
                sourceDocumentId,
                lifecycleStatus,
                timestampLabel);
            issues.AddRange(mapped.Issues);

            if (mapped.MasterRecord is null || mapped.DetailRecord is null)
            {
                continue;
            }

            var masterKey = (mapped.MasterRecord.SourceFamily, mapped.MasterRecord.SourceFamilyMasterId);
            if (!masters.TryGetValue(masterKey, out var existingMaster))
            {
                masters.Add(masterKey, mapped.MasterRecord);
            }
            else if (existingMaster == mapped.MasterRecord)
            {
                skippedDuplicateCount++;
            }
            else
            {
                issues.Add(new ParsedFactorPersistenceIssue(
                    "PARSED_FACTOR_PERSISTENCE_DUPLICATE_MASTER_CONFLICT",
                    "duplicate source-family master identity maps to different record content.",
                    $"Rows[{index}].source_family_master_id"));
            }

            var detailKey = (
                mapped.DetailRecord.SourceFamily,
                mapped.DetailRecord.SourceFamilyMasterId,
                mapped.DetailRecord.DetailExternalKey);
            if (!details.TryGetValue(detailKey, out var existingDetail))
            {
                details.Add(detailKey, mapped.DetailRecord);
            }
            else if (existingDetail == mapped.DetailRecord)
            {
                skippedDuplicateCount++;
            }
            else
            {
                issues.Add(new ParsedFactorPersistenceIssue(
                    "PARSED_FACTOR_PERSISTENCE_DUPLICATE_DETAIL_CONFLICT",
                    "duplicate factor identity maps to different detail record content.",
                    $"Rows[{index}].detail_external_key"));
            }
        }

        var command = new ParsedFactorPersistenceCommand(
            masters.Values,
            details.Values,
            skippedDuplicateCount,
            issues);
        var repositoryValidation = SourceFamilyRepositoryRegistry.ValidateInputs(
            "parsed_factor_persistence_command",
            command.MasterRecords,
            command.DetailRecords);

        if (repositoryValidation.Issues.Count == 0)
        {
            return command;
        }

        return new ParsedFactorPersistenceCommand(
            command.MasterRecords,
            command.DetailRecords,
            command.SkippedDuplicateCount,
            command.Issues.Concat(repositoryValidation.Issues.Select(FromRepositoryIssue)));
    }

    public static ParsedFactorPersistenceWriterResult Persist(
        ParserNormalizedOutputBatch parsedOutput,
        ISourceFamilyRepository repository,
        string? sourceDocumentId = null,
        string lifecycleStatus = "active",
        string timestampLabel = DefaultTimestampLabel)
    {
        var command = BuildCommand(parsedOutput, sourceDocumentId, lifecycleStatus, timestampLabel);
        if (command.Issues.Count > 0)
        {
            var status = IsOnlyNoRecords(command.Issues)
                ? ParsedFactorPersistenceStatus.NoRecords
                : ParsedFactorPersistenceStatus.FailedValidation;

            return new ParsedFactorPersistenceWriterResult(
                repository.ProviderName,
                status,
                command.MasterRecords.Count,
                command.DetailRecords.Count,
                0,
                0,
                command.SkippedDuplicateCount,
                command.Issues,
                command);
        }

        var repositoryResult = repository.PersistSourceFamilyRecords(
            command.MasterRecords,
            command.DetailRecords);
        var repositoryIssues = repositoryResult.Issues.Select(FromRepositoryIssue).ToArray();
        var resultStatus = repositoryResult.Status == SourceFamilyRepositoryPersistStatus.Declared
            ? ParsedFactorPersistenceStatus.Declared
            : ParsedFactorPersistenceStatus.FailedValidation;

        return new ParsedFactorPersistenceWriterResult(
            repositoryResult.ProviderName,
            resultStatus,
            command.MasterRecords.Count,
            command.DetailRecords.Count,
            repositoryResult.PersistedMasterCount,
            repositoryResult.PersistedDetailCount,
            command.SkippedDuplicateCount,
            repositoryIssues,
            command);
    }

    private static MappedRow MapRow(
        ParserNormalizedOutputRow row,
        int index,
        string? explicitSourceDocumentId,
        string lifecycleStatus,
        string timestampLabel)
    {
        if (!Enum.IsDefined(row.SourceFamily))
        {
            return new MappedRow(
                null,
                null,
                [
                    new ParsedFactorPersistenceIssue(
                        "PARSED_FACTOR_PERSISTENCE_UNSUPPORTED_SOURCE_FAMILY",
                        "source family must be GHG Protocol, DEFRA/DESNZ, or IPCC EFDB.",
                        $"Rows[{index}].SourceFamily"),
                ]);
        }

        var fields = row.Fields
            .GroupBy(field => field.Key, StringComparer.Ordinal)
            .ToDictionary(group => group.Key, group => group.First().Value, StringComparer.Ordinal);
        var issues = new List<ParsedFactorPersistenceIssue>();
        var resolvedSourceDocumentId = ResolveSourceDocumentId(row, fields, explicitSourceDocumentId);
        var factorValue = TextOrNull(Field(fields, "factor_value", "value"));
        var factorUnit = TextOrNull(Field(fields, "factor_unit", "unit"));

        var requiredValues = new Dictionary<string, string?>
        {
            ["source_document_id"] = resolvedSourceDocumentId,
            ["factor_value"] = factorValue,
            ["factor_unit"] = factorUnit,
        };
        foreach (var pair in requiredValues)
        {
            if (pair.Value is null)
            {
                issues.Add(new ParsedFactorPersistenceIssue(
                    "PARSED_FACTOR_PERSISTENCE_MISSING_REQUIRED_FIELD",
                    "parsed factor persistence requires a non-empty value.",
                    $"Rows[{index}].{pair.Key}"));
            }
        }

        if (issues.Count > 0)
        {
            return new MappedRow(null, null, issues);
        }

        var masterExternalKey = TextOrNull(Field(fields, "master_external_key")) ?? DefaultMasterExternalKey(row, fields);
        var detailExternalKey = TextOrNull(Field(fields, "detail_external_key")) ?? DefaultDetailExternalKey(row, fields);
        var familyPrefix = row.SourceFamily switch
        {
            SourceFamily.GhgProtocol => "ghg",
            SourceFamily.DefraDesnz => "defra",
            SourceFamily.IpccEfdb => "ipcc",
            _ => row.SourceFamily.ToWireName(),
        };
        var masterId = TextOrNull(Field(fields, "source_family_master_id"))
            ?? $"{familyPrefix}_master_{StableDigest(familyPrefix, masterExternalKey)[..16]}";
        var detailId = TextOrNull(Field(fields, "source_family_detail_id"))
            ?? $"{familyPrefix}_detail_{StableDigest(familyPrefix, masterId, detailExternalKey)[..16]}";

        var masterRecord = new SourceFamilyMasterRecord(
            row.SourceFamily,
            masterId,
            resolvedSourceDocumentId!,
            masterExternalKey,
            lifecycleStatus,
            TextOrNull(Field(fields, "effective_from")),
            TextOrNull(Field(fields, "effective_to")),
            StableDigest("master", familyPrefix, resolvedSourceDocumentId!, masterExternalKey, lifecycleStatus),
            timestampLabel,
            timestampLabel);
        var detailRecord = new SourceFamilyDetailRecord(
            row.SourceFamily,
            detailId,
            masterId,
            detailExternalKey,
            factorValue!,
            factorUnit!,
            lifecycleStatus,
            StableDigest("detail", familyPrefix, masterId, detailExternalKey, factorValue!, factorUnit!),
            timestampLabel,
            timestampLabel);

        return new MappedRow(masterRecord, detailRecord, issues);
    }

    private static void AppendRowValidationIssues(
        ParserNormalizedOutputRow row,
        int index,
        ICollection<ParsedFactorPersistenceIssue> issues)
    {
        foreach (var error in row.Validate().Errors)
        {
            issues.Add(new ParsedFactorPersistenceIssue(
                "PARSED_FACTOR_PERSISTENCE_INVALID_NORMALIZED_ROW",
                error,
                $"Rows[{index}]"));
        }
    }

    private static string? ResolveSourceDocumentId(
        ParserNormalizedOutputRow row,
        IReadOnlyDictionary<string, string?> fields,
        string? explicitSourceDocumentId)
    {
        var explicitValue = TextOrNull(explicitSourceDocumentId);
        if (explicitValue is not null)
        {
            return explicitValue;
        }

        var fieldValue = TextOrNull(Field(fields, "source_document_id"));
        if (fieldValue is not null)
        {
            return fieldValue;
        }

        var artifactReference = TextOrNull(Field(fields, "provenance_artifact_reference", "artifact_reference"))
            ?? TextOrNull(row.ArtifactReference);
        var checksum = TextOrNull(Field(fields, "provenance_checksum_value", "source_checksum_sha256"));
        if (artifactReference is null && checksum is null)
        {
            return null;
        }

        return $"source_document_{StableDigest(row.SourceFamily.ToWireName(), row.SourceKey, artifactReference, checksum)[..24]}";
    }

    private static string DefaultMasterExternalKey(
        ParserNormalizedOutputRow row,
        IReadOnlyDictionary<string, string?> fields)
    {
        var sourceYear = TextOrNull(Field(fields, "source_year")) ?? "unknown-year";
        var sourceVersion = TextOrNull(Field(fields, "source_version")) ?? "unknown-version";
        var factorId = TextOrNull(Field(fields, "factor_id")) ?? row.RowIdentifier;

        return $"{sourceYear}:{sourceVersion}:{factorId}";
    }

    private static string DefaultDetailExternalKey(
        ParserNormalizedOutputRow row,
        IReadOnlyDictionary<string, string?> fields)
    {
        var factorId = TextOrNull(Field(fields, "factor_id")) ?? row.RowIdentifier;
        var factorUnit = TextOrNull(Field(fields, "factor_unit", "unit")) ?? "unknown-unit";
        var gas = TextOrNull(Field(fields, "greenhouse_gas", "gas"));

        return gas is null
            ? $"{factorId}:{factorUnit}"
            : $"{factorId}:{factorUnit}:{gas}";
    }

    private static string? Field(IReadOnlyDictionary<string, string?> fields, params string[] names)
    {
        foreach (var name in names)
        {
            if (fields.TryGetValue(name, out var value))
            {
                return value;
            }
        }

        return null;
    }

    private static string? TextOrNull(string? value)
    {
        var text = value?.Trim();

        return string.IsNullOrEmpty(text) ? null : text;
    }

    private static string StableDigest(params string?[] values)
    {
        var payload = JsonSerializer.Serialize(values);

        return Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(payload))).ToLowerInvariant();
    }

    private static ParsedFactorPersistenceIssue FromRepositoryIssue(SourceFamilyRepositoryIssue issue) =>
        new(issue.Code, issue.Message, issue.FieldName, issue.Severity);

    private static bool IsOnlyNoRecords(IReadOnlyList<ParsedFactorPersistenceIssue> issues) =>
        issues.Count == 1 && issues[0].Code == "PARSED_FACTOR_PERSISTENCE_NO_RECORDS";

    private sealed record MappedRow(
        SourceFamilyMasterRecord? MasterRecord,
        SourceFamilyDetailRecord? DetailRecord,
        IEnumerable<ParsedFactorPersistenceIssue> Issues);
}
